File: src/main/java/com/pubnub/api/services/PushService.java
package com.pubnub.api.services;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.QueryMap;

import java.util.List;
import java.util.Map;

public interface PushService {

    @GET("v1/push/sub-key/{subKey}/devices/{pushToken}")
    Call<List<Object>> modifyChannelsForDevice(@Path("subKey") String subKey,
                                               @Path("pushToken") String pushToken,
                                               @QueryMap Map<String, String> options);

    @GET("v1/push/sub-key/{subKey}/devices/{pushToken}/remove")
    Call<List<Object>> removeAllChannelsForDevice(@Path("subKey") String subKey,
                                                  @Path("pushToken") String pushToken,
                                                  @QueryMap Map<String, String> options);

    @GET("v1/push/sub-key/{subKey}/devices/{pushToken}")
    Call<List<String>> listChannelsForDevice(@Path("subKey") String subKey,
                                             @Path("pushToken") String pushToken,
                                             @QueryMap Map<String, String> options);

    // V2 (APNS2)

    @GET("v2/push/sub-key/{subKey}/devices-apns2/{deviceApns2}")
    Call<List<Object>> modifyChannelsForDeviceApns2(@Path("subKey") String subKey,
                                                    @Path("deviceApns2") String deviceApns2,
                                                    @QueryMap Map<String, String> options);

    @GET("v2/push/sub-key/{subKey}/devices-apns2/{deviceApns2}")
    Call<List<String>> listChannelsForDeviceApns2(@Path("subKey") String subKey,
                                                  @Path("deviceApns2") String deviceApns2,
                                                  @QueryMap Map<String, String> options);

    @GET("v2/push/sub-key/{subKey}/devices-apns2/{deviceApns2}/remove")
    Call<List<Object>> removeAllChannelsForDeviceApns2(@Path("subKey") String subKey,
                                                       @Path("deviceApns2") String deviceApns2,
                                                       @QueryMap Map<String, String> options);


}

File: core/src/main/kotlin/com/pubnub/api/services/PushService.kt
package com.pubnub.api.services

import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.QueryMap

internal interface PushService {

    @GET("v1/push/sub-key/{subKey}/devices/{pushToken}")
    fun modifyChannelsForDevice(
        @Path("subKey") subKey: String,
        @Path("pushToken") pushToken: String,
        @QueryMap options: Map<String, String>
    ): Call<Void>

    @GET("v1/push/sub-key/{subKey}/devices/{pushToken}/remove")
    fun removeAllChannelsForDevice(
        @Path("subKey") subKey: String,
        @Path("pushToken") pushToken: String,
        @QueryMap options: Map<String, String>
    ): Call<Void>

    @GET("v1/push/sub-key/{subKey}/devices/{pushToken}")
    fun listChannelsForDevice(
        @Path("subKey") subKey: String,
        @Path("pushToken") pushToken: String,
        @QueryMap options: Map<String, String>
    ): Call<List<String>>

    // V2 (APNS2)

    // V2 (APNS2)
    @GET("v2/push/sub-key/{subKey}/devices-apns2/{deviceApns2}")
    fun modifyChannelsForDeviceApns2(
        @Path("subKey") subKey: String,
        @Path("deviceApns2") deviceApns2: String,
        @QueryMap options: Map<String, String>
    ): Call<Void>

    @GET("v2/push/sub-key/{subKey}/devices-apns2/{deviceApns2}")
    fun listChannelsForDeviceApns2(
        @Path("subKey") subKey: String,
        @Path("deviceApns2") deviceApns2: String,
        @QueryMap options: Map<String, String>
    ): Call<List<String>>

    @GET("v2/push/sub-key/{subKey}/devices-apns2/{deviceApns2}/remove")
    fun removeAllChannelsForDeviceApns2(
        @Path("subKey") subKey: String,
        @Path("deviceApns2") deviceApns2: String,
        @QueryMap options: Map<String, String>
    ): Call<Void>
}
